src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -647,7 +647,7 @@ private void ProduceTagsSynchronously(TaggerContext<TTag> context)
                 {
                     foreach (var (latestBuffer, latestSpans) in newTagTrees)
                     {
-                        var snapshot = spansToTag.First(s => s.SnapshotSpan.Snapshot.TextBuffer == latestBuffer).SnapshotSpan.Snapshot;
+                        var snapshot = spansToTag.FirstOrDefault(s => s.SnapshotSpan.Snapshot.TextBuffer == latestBuffer).SnapshotSpan.Snapshot ?? latestBuffer.CurrentSnapshot;
```

Comment:
this is extremely scary to me.  I would really want to understnad when/how could this happen.  How do not have any spans we want to tag that match the buffer of the we just generated?  that seems like this is masking a bug happening elsehwere...

Reply:
I agree. At first I pursued two other approaches:

1. Avoiding this method if `spansToTag` is empty (turns out it was not empty)
2. Using `cancellationToken.ThrowIfCancellationRequested` if no matching spans were found (cancellation was not requested)

I was not able to determine how/why we enter this state, but it's causing a spike in Watson since last week and our integration tests are consistently failing.

Reply:
can we NFW at least.  this feels bad to me.  

Note: a better approach might be to have this data just passed through the API instead of trying to re-establish it at this point in the code.  It might make it clearer at which point the info is getting 'out of sync' as it were...

Reply:
Is this already creating fatal exceptions? Do we have dumps for those?

Reply:
But I agree with @CyrusNajmabadi -- this is masking a very bad problem. Let's only do this if we have a clear plan in place for how to address it further.

Reply:
I am +1 on not crashing. I don't believe most of tagger issue is serious enough to crash VS. I believe NFW should be fine for such issue.

Reply:
we should split the issue 2. first don't carsh. second actually fix the root cause.

Reply:
> Is this already creating fatal exceptions? Do we have dumps for those?

The Watson bucket is listed in the original comment.

Reply:
@heejaechang Given the likelihood of users hitting this, I'm not comfortable with NFW. Watson is showing spikes even though this code is in extremely limited distribution. However, we could revert the change that started this (#26512) until the investigation is complete.

Reply:
@heejaechang Unless we have a way to *rate limit* the NFWs. That seems fine. I just don't want it to be a case where many of them are getting reported and impacting someone's experience.

Reply:
@sharwell NFW should not do that. if it does, we should improve NFW rather than not using it. can you actually try NFW and see whether it actually cause user experience go to bad. sure when it send Watson it will slow down VS. but can you see whether 1. NFW send it repeatedly and 2. NFW send same dump from same users from multiple different session?

we should use VS APIs that everyone else is also using.

Reply:
Let's not turn this into a rehash of our NFW policy here; we have actual bugs to fix.

Reply:
@CyrusNajmabadi @sharwell Looking at the code a bit, I've got a theory.

1. This is happening around code that seems to be dealing with tagging multiple buffers
2. We're seeing crashes in our interactive tests.

My guess is this is the "special" logic we have around interactive tagging of reference highlighting. Basically, we tag multiple buffers at once and keep all the tags for them. The one thing I don't know in this system is if a buffer goes away, who removes the cached tags for that buffer? Or would newTagTrees still have tags for that older buffer, which would trip up this code?

Reply:
@jasonmalinowski i highly suspect your'e right.  This is why i proposed the above:

> Note: a better approach might be to have this data just passed through the API instead of trying to re-establish it at this point in the code. It might make it clearer at which point the info is getting 'out of sync' as it were...

It's pretty clear to me that this code is trying to synthesize a relation that doesn't seem to exist.  So, it would likely be better if the entire codepath was explicit about passing this data along, and then had explicit code to try to decide what to do if/when the data didn't meet its expectation.  i.e. if the buffer we have for "newTreeTags" doesn't match teh one for "oldTreeTags", we really need to have explicit code to understand what that means, and what to do about it.

Note: this is also pretty scary to me given that taggers are supposed to be either view or buffer associated.  If we're breaking that assumption... then we need to do some pretty substantial rethinking of things.  

As an alternative, we could try to figure out why we're getting a tagger that isn't really associated with a single buffer, and we could try to change things to make it so that we have that invariant.  i.e. instead of reusing this tagger across buffer, ensure we have a unique tagger per buffer.  That way we coudl then put in invariant-checks in the code that we weren't operating on the wrong buffer.

Reply:
> Note: this is also pretty scary to me given that taggers are supposed to be either view or buffer associated. If we're breaking that assumption... then we need to do some pretty substantial rethinking of things.

The special case here is the multi-buffer tagging we do for reference highlighting: we have a single one of these for an entire view, where we're tagging each of the subject buffers in the view. This was the explicit new design when we did this refactoring originally.

Reply:
Going back to the "why?" here. Reference highlighting in interactive is tricky. It can't be a traditional buffer tagger, because you might be tagging a buffer (i.e. a submission) but the caret can be in _another_ buffer. If we make it a view buffer that tags the _surface_ buffer, you end up with other interestingness:

1. This means our tagger is tagging the surface buffer directly, and who knows what content type that is.
2. We now get to sort out what is going on with projection buffers as edits happen and make sure we are still giving sane tags.

Reply:
Ok.  if it's explicitly intentional that we have a single tagger for a collection of buffers, then it feels to me liek we then need to make the internal data-types more explicit about what they are assocaited with.  i.e. instead of just getting a random set of tags, you get back a set of tags-for-a-particular-buffer.  And the entire stack here is designed around appropriately query and managing the information on a per-buffer level.